### PR TITLE
Do not look for block variants, if not supporting block-templates

### PR DESCRIPTION
**Pull request by spacedmonkey**
## What?
Do not look for block variants if theme does not support them. Saving a database query.

Original ticket [here](https://core.trac.wordpress.org/ticket/56923).

### CODE REVIEW

**Comment by t-hamano** [comment]
I don't know the details of how it works, but I am concerned that it may affect the classic theme that supports template parts. In the Classic theme, opt-in the Template Parts Editor as follows:

```php
function add_block_template_part_support() {
    add_theme_support( 'block-template-parts' );
}
add_action( 'after_setup_theme', 'add_block_template_part_support' );
```

Then, when we try to load other template parts within the template parts editor, the behavior changes. 

### On trunk

trunk

### On this branch

pr

**Review by peterwilsoncc — DISMISSED** [review]
This looks good to me based on manual testing.

**Note:**

On my Gutenberg development box, I needed to modify the equivelant path in WordPress, `/wp-includes/blocks/template-part.php`, to confirm this change.

For whatever reason, the plugin wasn't using its version of the PHP files.

**Comment by peterwilsoncc** [comment]
Having dug in to this following @t-hamano 's comment above, I think it needs to return early with the following:

```php
if (
	! current_theme_supports( 'block-templates' )
	&& ! current_theme_supports( 'block-template-parts' )
) {
	return [];
}
```

`block-templates` support implies `block-template-parts` support but this doesn't appear to be reflected when calling `current_theme_supports()`.

**Comment by talldan** [comment]
This is a performance issue? The ticket/PR is a bit light on details.

This idea behind this PR seems like an ok patch for a short term improvement in 6.1.1, but doesn't really get to the core of the issue. Other code can still call `get_block_templates` and reintroduce the problem.

The two questions I have:
- Why does `get_block_templates` go as far as making a DB query if templates/template parts are not supported? Could the early return be added to that function?
- Why is the template part block registered for themes that don't support template parts? Not registering it would also avoid this code being triggered and avoid a bunch of unnecessary work.

It might be good to look at a different fix for 6.2.

**Comment by Mamaduka** [comment]
> block-templates support implies block-template-parts support

@peterwilsoncc, I don't think that's true. Those are two separate "flags" for classic themes.

**Comment by peterwilsoncc** [comment]
@Mamaduka This is what I see in TT3, which includes template parts but doesn't indicate `block-template-parts` support:

```
$ wp shell
wp> get_stylesheet_uri();
=> string(69) "http://vagrant.local/wp/wp-content/themes/twentytwentythree/style.css"
wp> current_theme_supports( 'block-templates' )
=> bool(true)
wp> current_theme_supports( 'block-template-parts' )
=> bool(false) 
```

Maybe there is some nuance that I am missing re the purpose the theme support flags serve.

@talldan Yeah, a perf issue to avoid unneeded DB queries. Thanks for tagging appropriately.

I think you're right, there is room to improve further in 6.2 by basically nooping any of the database queries template blocks make.

**Comment by Mamaduka** [comment]
@peterwilsoncc, the `block-template-parts` is a new flag that partially enables Site Editor for classic themes. Since block themes already have full access to Site Editor, there was no need to set the theme support for them.

Internally the flag controls what pieces of the UI are hidden/inaccessible.

Here's a dev note for more details: https://make.wordpress.org/core/2022/10/04/block-based-template-parts-in-traditional-themes/.

File: packages/block-library/src/template-part/index.php [hunk 1]
```diff
@@ -193,6 +193,11 @@ function build_template_part_block_instance_variations() {
 	if ( wp_installing() ) {
 		return array();
 	}
+
+	if ( ! current_theme_supports( 'block-templates' ) ) {
```
Comment (spacedmonkey) [inline packages/block-library/src/template-part/index.php]:
```suggestion
	if ( ! current_theme_supports( 'block-templates' ) && ! current_theme_supports(  'block-template-parts' ) ) {
```

**Comment by spacedmonkey** [comment]
Can you pull and tests again @t-hamano

**Comment by t-hamano** [comment]
I have checked the behavior again and it is working as expected.

I also checked again the support variations in the classic theme, and I believe they are as follows:

### Support `block-templates` only

We can use the template editor in the post editor. The template parts editor menu is not accessible. However, we can call template parts defined in the theme from within the template editor.

### Support `block-template-parts` only

We cannot use the template editor in the post editor. we can access ****the template parts editor menu**,** and in the template parts editor we can also call template parts defined in your theme.

### Support `block-templates` and `block-template-parts`

We can use the template editor in the post editor. We can access the template parts editor menu, and in the template parts editor, we can also call template parts defined in the theme.


Therefore, I think it would be appropriate not to query only for template parts unless both `block-templates` and `block-template-parts` are supported.

File: packages/block-library/src/template-part/index.php [hunk 2]
```diff
@@ -193,6 +193,11 @@ function build_template_part_block_instance_variations() {
 	if ( wp_installing() ) {
 		return array();
 	}
+
+	if ( ! current_theme_supports( 'block-templates' ) && ! current_theme_supports(  'block-template-parts' ) ) {
```
Comment (t-hamano) [inline packages/block-library/src/template-part/index.php]:
```suggestion
	if ( ! current_theme_supports( 'block-templates' ) && ! current_theme_supports( 'block-template-parts' ) ) {
```

To pass lint check

**Comment by spacedmonkey** [comment]
I guess the question is, do we even need to register this block on themes that do not support them anyway. Seems like it would never be used.

**Comment by spacedmonkey** [comment]
I would love some context here from @talldan.

**Review by Mamaduka — APPROVED** [review]
The changes here look good to me. Thank you, @spacedmonkey!

I agree with @talldan that we should explore a more general solution for WP 6.2.

Even for the block themes, variations are only needed by the editors. There's no use in querying them when a user visits a page.

One option we can explore for the next major release is the lazy initialization of block variables.

**Comment by spacedmonkey** [comment]
> One option we can explore for the next major release is the lazy initialization of block variables.

Created a ticket https://github.com/WordPress/gutenberg/issues/45601.

**Comment by sybrew** [comment]
Could the concerns of https://github.com/WordPress/gutenberg/pull/45362#issuecomment-1294428400 above also be addressed, please? 

I think no database queries for theme support, and especially not anything via WP Query, should be performed before a template can be determined (`template_redirect`).

**Pull request merged**